API Move IterativeImputer outside from experimental (#34214)
scikit-learn/scikit-learn: doc/modules/impute.rst

.. currentmodule:: sklearn.impute

For various reasons, many real-world datasets contain missing values, often
encoded as blanks, NaNs or other placeholders. They arise from faulty
measurements, unanswered questionnaire items, or information that was simply
never recorded, and can affect both the data ``X`` and the target ``y``. Most
scikit-learn estimators assume that every entry of an array is numerical and
holds meaning, and therefore cannot be trained directly on incomplete data.

A naive way to meet that requirement is to discard every row or column that
contains a missing value. This is detrimental on two counts: it throws away

> potentially valuable information, and it generally introduces bias, since the

Reviewer: ```suggestion
contains a missing value. This is detrimental on two counts: valuable 
information from those cases is lost, and it generally introduces bias, since the
```
Nit.

remaining samples are rarely representative of the original population (unless
values are missing completely at random). The same caution applies to the
target: silently dropping samples whose outcome ``y`` is unknown biases the
analysis. In particular, when an outcome has not been observed yet, the problem

> is one of censoring and should be handled with dedicated methods from `survival

Reviewer: ```suggestion
analysis. In particular, when an outcome has not been observed yet, it is a
censoring problem and should be handled with dedicated methods from `survival
```
Slight wording fix. But mentioning "censoring" is a bit too much survival analysis jargon, I think. I don't know what it means; many people won't either. Can we express that in more general terms?

Author: I think it is good to mention the jargon but between brackets we can mention an example to illustrate what it is.

strategy than discarding data is to impute the missing values, i.e. to infer
them from the known part of the data (see the glossary entry on
:term:`imputation`). scikit-learn provides simple column statistics with
:class:`SimpleImputer`, as well as the model-based :class:`IterativeImputer`
and :class:`KNNImputer`.

When the goal is prediction rather than reconstructing the data, a few
high-level guidelines help choose among these tools [3]_:

- Start simple: constant or mean / most-frequent imputation with
  :class:`SimpleImputer` is a strong and cheap baseline, and more elaborate
  imputation often brings only marginal gains in predictive performance.
- Flag missing entries: adding a missingness indicator (the ``add_indicator``
  option of the imputers, or :class:`MissingIndicator`) tends to help
  prediction, even when values are missing completely at random.
- Prefer expressive models: flexible estimators benefit even less from
  sophisticated imputation, and some handle missing values natively without
  any imputation (see :ref:`estimators_that_handle_nan`).
- Invest in imputation quality mainly when reconstructing the data itself, not

>   prediction, is the objective.

Reviewer: ```suggestion

Invest in imputation quality mainly when reconstructing the data itself, not
prediction, is the objective.
```
Since this doesn't fit with the "When the goal is prediction rather than reconstructing the data, a few" introduction of these bullet points, maybe take it out.

Reviewer 2: I actually think that it is OK to have it here, as it will be useful for a bunch of people

Author: OK so I put back the statement.

.. topic:: Key points for machine-learning with missing values

   * Avoid dropping rows with missing values; it risks creating bias.

   * Tree-based learners can natively predict on data with missing values
     and work fairly well with no additional cost.

  * Imputation can be very computationaly costly, and quickly hits
    diminishing returns to improve subsequent prediction performance [3]_.

For various reasons, many real-world datasets contain missing values, often
encoded as blanks, NaNs or other placeholders. They arise from faulty
measurements, unanswered questionnaire items, or information that was simply
never recorded, and can affect both the data ``X`` and the target ``y``. Most
scikit-learn estimators assume that every entry of an array is numerical and
holds meaning, and therefore cannot be trained directly on incomplete data.

A naive way to meet that requirement is to discard every row or column that
contains a missing value. This is detrimental on two counts: valuable 
information from those cases is lost, and it generally introduces bias, since the
remaining samples are rarely representative of the original population (unless

> values are missing completely at random). The same caution applies to the
> target: silently dropping samples whose outcome ``y`` is unknown biases the
> analysis. In particular, when an outcome has not been observed yet, it is a
> censoring problem and should be handled with dedicated methods from `survival
> analysis <https://en.wikipedia.org/wiki/Survival_analysis>`_ rather than by
> discarding data.

Author: ```suggestion
analysis. In particular, when an outcome has not been observed yet, it is a
`censoring
problem <https://en.wikipedia.org/wiki/Survival_analysis#Censoring>`_
and should be handled with dedicated methods from `survival
analysis <https://en.wikipedia.org/wiki/Survival_analysis>`_ rather than by
discarding data.
```

.. currentmodule:: sklearn.impute

.. topic:: Key points for machine-learning with missing values

   * Avoid dropping rows with missing values; it risks creating bias.

>    * Tree-based learners can natively predict on data with missing values

Reviewer 2: ```suggestion
   * :ref:`Some supervised learning methods <estimators_that_handle_nan>` (typically
     tree-based learners) can natively predict on data with missing values
```

A naive way to meet that requirement is to discard every row or column that
contains a missing value. This is detrimental on two counts: valuable 
information from those cases is lost, and it generally introduces bias, since the
remaining samples are rarely representative of the original population (unless
values are missing completely at random). The same caution applies to the
target: silently dropping samples whose outcome ``y`` is unknown biases the
analysis. In particular, when an outcome has not been observed yet, it is a
`censoring
problem <https://en.wikipedia.org/wiki/Survival_analysis#Censoring>`_
and should be handled with dedicated methods from `survival
analysis <https://en.wikipedia.org/wiki/Survival_analysis>`_ rather than by
discarding data.

The tools described on this page address missing values in ``X``. A better
strategy than discarding data is to impute the missing values, i.e. to infer
them from the known part of the data (see the glossary entry on
:term:`imputation`). scikit-learn provides simple column statistics with
:class:`SimpleImputer`, as well as the model-based :class:`IterativeImputer`
and :class:`KNNImputer`.

> 

Author: ```suggestion

Invest in imputation quality mainly when reconstructing the data itself, not
prediction, is the objective.
```

Pipeline as a way to build a composite estimator that supports imputation.
See :ref:`sphx_glr_auto_examples_impute_plot_missing_values.py`.

.. _iterative_imputer_convergence:

Convergence and diminishing returns
-----------------------------------

:class:`IterativeImputer` repeats the round-robin imputation for ``max_iter``
rounds and stops early when the change between two consecutive rounds falls
below ``tol`` (when ``sample_posterior=False``). In practice, the imputed
values often do not converge: the round-robin scheme is not guaranteed to
reach a fixed point, so the number of iterations is best seen as a trade-off
against the available time budget rather than a target to be met.

This is rarely a problem when the goal is prediction. Le Morvan and Varoquaux
[3]_ report strongly diminishing returns: better imputation accuracy yields
only small gains in downstream predictive performance, especially with an

> expressive model and a missingness indicator (see ``add_indicator``).

Reviewer: ```suggestion
expressive model and a missingness indicator (``add_indicator``).
```
Otherwise, I'd expect a clickable link.

In practice, prefer a small, fixed ``max_iter`` together with a missingness
indicator and an expressive downstream model over investing in convergence.
When the goal is to *reconstruct* the data rather than predict, the number of
iterations and the choice of imputer matter more and should be assessed on the

> task at hand.

Reviewer: ```suggestion
In practice, prefer a small, fixed ``max_iter`` (i.e. `max_iter=10`) together with 
a missingness indicator and an expressive downstream model over investing 
in convergence. When imputation itself is the goal (e.i. for *reconstructing data*) 
rather than predicting, the number of iterations (i.e. `max_iter>50`) and the 
choice of imputer matter more and should be assessed on the task at hand.
```
Let's advise people what "small" means. 10 repetitions are used for prediction in the paper and 50 for reconstruction, so maybe suggest these?

Also, I find the reference to reconstruction a bit too optimistic and turned it into an example usage.

Reviewer 2: +1 with suggestion 10 for prediction and 50 for reconstruction. Thanks for bringing this up @StefanieSenger
